src/url-params.ts

```diff
@@ -0,0 +1,105 @@
+import {
+	getCookie,
+	hasCookieSupport,
+	setCookie,
+} from '@analytics/cookie-utils';
+import union from 'lodash/union';
+import { Mixpanel } from 'mixpanel-browser';
+
+const URL_PARAM_DEVICE_ID = 'd_id';
+const COOKIES_DEVICE_IDS = '__analytics_dids';
+
+const deviceIdSeparator = /\s*,\s*/;
+
+/**
+ * AnalyticsUrlParams helps with handling analytics-related URL parameters
+ * like anonymous device identifier.
+ */
+export class AnalyticsUrlParams {
+	private readonly cookiesSupported = hasCookieSupport();
+	private deviceIds: string[];
+
+	constructor(private mixpanel?: Mixpanel) {
+		const storedValue = this.cookiesSupported
+			? getCookie(COOKIES_DEVICE_IDS)
+			: null;
+		this.setDeviceIds(storedValue, null);
+	}
+
+	private setDeviceIds(
+		inputIdString: string | null,
+		currentDeviceId: string | null,
+	) {
+		const list = inputIdString ? inputIdString.split(deviceIdSeparator) : [];
+		if (currentDeviceId) {
+			list.push(currentDeviceId);
+		}
+		this.deviceIds = union(this.deviceIds, list);
+		if (this.cookiesSupported) {
+			setCookie(COOKIES_DEVICE_IDS, this.deviceIds.join(','));
+		}
+		return list.length > 0 ? list[0] : null;
+	}
+
+	clearCookies() {
+		if (this.cookiesSupported) {
+			setCookie(COOKIES_DEVICE_IDS, '');
+		}
+	}
+
+	/**
+	 * Analyzes the query string and stores the anonymous device ID if it's there.
+	 * Returns null if the current URL query string should not be modified after consuming the data.
+	 *
+	 * @param queryString URL query string to process
+	 * @return a modified query string that hides processed/consumed parameters if the query string should be modified
+	 */
+	consumeUrlParameters(queryString: string): string | null {
+		const params = new URLSearchParams(queryString);
+
+		const passedDeviceId = params.get(URL_PARAM_DEVICE_ID);
+		if (passedDeviceId) {
+			const originalMixpanelId =
+				this.mixpanel != null ? this.mixpanel.get_distinct_id() : null;
+
+			const newCurrentDeviceId = this.setDeviceIds(
+				passedDeviceId,
+				originalMixpanelId,
+			);
+
+			if (this.mixpanel != null && newCurrentDeviceId) {
+				// Switch mixpanel ID to using the passed device ID, so we can track events in one timeline branch.
+				// Previous user activity recorded with another device ID will be merged upon signup,
+				// since we store the previous ID to pass it to other sites.
+				this.mixpanel.register({
+					distinct_id: newCurrentDeviceId,
+					$device_id: newCurrentDeviceId,
+				});
+			}
+
+			params.delete(URL_PARAM_DEVICE_ID);
+			return params.toString();
+		} else {
+			return null;
+		}
+	}
+
+	/**
+	 * @return all anonymous device IDs that can be passed to other sites
+	 */
+	allDeviceIds() {
+		const mixpanelId = this.mixpanel ? [this.mixpanel.get_distinct_id()] : [];
```

Comment:
I think we can also do this check and assignment in the constructor, since that's where we get mixpanel passed. In that case we can have this method to be `getAllDeviceIds` and it will simply return `this.deviceIds`.

Reply:
See https://github.com/balena-io/analytics-client/pull/1/commits/bc5b491cad22dd7a8de7498a68b5772046bf4285

What you describe was my initial implementation, but then I changed it in the commit above making sure we don't expect mixpanel library to be initialized in the constructor.

Reply:
If mixpanel is not passed in the constructor, there is no way to update it afterwards, as it is private. If a mixpanel client is indeed passed to the constructor, but it is not ready yet, then `this.mixpanel` will always return true, so this check will be wrong. Maybe I am missing something?

Reply:
A typical use of mixpanel:
```
import mixpanel from 'mixpanel-browser';

// AnalyticsUrlParams can be initialized at this moment of time.

mixpanel.init(token, options);

// Only now you can
mixpanel.get_distinct_id();
```

So we can do `const analyticsTool = new AnalyticsUrlParams(mixpanel)`, but `init ` is not yet called on the mixpanel object.

Reply:
But `mixpanel` is never falsy, so `this.mixpanel ? ... : ...` will always be true.

Reply:
The constructor is declared in a way it allows the mixpanel object to be skipped.
The logic handles this particular case (so it works when we do `new AnalyticsUrlParams()`).
The main point is that `get_distinct_id` call should be delayed, which is why `allDeviceIds` method is a bit more complicated.

Speaking about `isInitialized`, there is [no such method](https://github.com/mixpanel/mixpanel-js/issues/133) in mixpanel.